SerialPrograms/Source/CommonFramework/Audio/AudioDisplayWidget.cpp

```diff
@@ -0,0 +1,564 @@
+/*  Audio Display Widget
+ *
+ *  From: https://github.com/PokemonAutomation/Arduino-Source
+ *
+ */
+
+#include "AudioDisplayWidget.h"
+#include "AudioInfo.h"
+#include "CommonFramework/Logging/Logger.h"
+#include "Kernels/AbsFFT/Kernels_AbsFFT.h"
+
+#include <QVBoxLayout>
+#include <QLabel>
+#include <QLinearGradient>
+#include <QResizeEvent>
+#include <QPaintEvent>
+#include <QPainter>
+#include <QColor>
+#include <QIODevice>
+#include <QBuffer>
+
+#include <iostream>
+#include <cfloat>
+#include <chrono>
+
+// #define USE_FFTREAL
+
+#ifdef USE_FFTREAL
+#include <fftreal_wrapper.h>
+#else
+const int FFTLengthPowerOfTwo = 14;
+#endif
+
+const int NUM_FFT_SAMPLES = 1 << FFTLengthPowerOfTwo;
+const int NUM_FFT_WINDOWS = 20;
+const int FFT_SLIDING_WINDOW_STEP = NUM_FFT_SAMPLES/4;
+
+#if QT_VERSION_MAJOR == 5
+#include <QAudioDeviceInfo>
+#include <QAudioInput>
+#include <QAudioOutput>
+#include <QtEndian>
+
+// In Qt5 there is no function to normalize sample value.
+// So here I have to write a new one using
+// the implementation in Qt6 here from https://github.com/sailfishos-mirror/qtmultimedia
+// and Qt5 example code from https://code.qt.io/cgit/qt/qtmultimedia.git/tree/examples/multimedia/audioinput/audioinput.cpp?h=5.15
+float normalizeSampleValue(const QAudioFormat& format, const void *ptr){

```

Comment:
ugh... lol

Reply:
I am a bit worried about the performance of this function if it's being called on every single sample. But we'll come back to that later.